### PR TITLE
Remove Dispose(disposing) from objects without finalization

### PR DESCRIPTION
**Pull request by tmat**
Follow up on feedback on https://github.com/dotnet/roslyn/pull/40395.

The rules on which dispose patterns to use:

The base type that implements IDisposable chooses whether to implement the legacy pattern (with `protected virtual void Dispose(bool)`) or the new pattern (with `public virtual void Dispose()`)
- A type using the new pattern must not add a finalizer
- A type derived from a type that uses the new pattern must not add a finalizer

### CODE REVIEW

**Comment by tmat** [comment]
@sharwell

**Comment by CyrusNajmabadi** [comment]
> A type using the new pattern must not add a finalizer

Even a debug-only finalizer that validates that someone didn't forget to call `.Dispose`?  These are *great* for catching bugs and I would def still want to use those.

**Comment by tmat** [comment]
@CyrusNajmabadi Those debug only finalizers don't really do resource cleanup, so they can be used with the new pattern. I'd say such finalizer needs to be in the base class. We have one here: https://github.com/dotnet/roslyn/pull/40782/files#diff-51efd8b513355aaf023221a46dfacca3R239

**Comment by CyrusNajmabadi** [comment]
Great.  WFM then.  Thanks! :)

File: src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs [hunk 1]
```diff
@@ -220,7 +215,7 @@ protected Connection()
             public abstract Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object> arguments, CancellationToken cancellationToken);
             public abstract Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync, CancellationToken cancellationToken);
 
-            protected virtual void Dispose(bool disposing)
+            protected virtual void DisposeImpl()
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs]: 💭 Even this one seems like it would benefit from just using `Dispose()`

Comment (tmat) [reply src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs]: We have a pattern where the base class only calls DisposeImpl if the object has not been disposed yet.

**Pull request merged**